uPortal-security/uPortal-security-permissions/src/main/java/org/apereo/portal/security/provider/AuthorizationImpl.java

```diff
@@ -571,22 +587,54 @@ public boolean doesPrincipalHavePermission(
         final IPermissionOwner ipOwner = permissionOwnerDao.getPermissionOwner(owner);
         final IPermissionActivity ipActivity =
                 permissionOwnerDao.getPermissionActivity(owner, activity);
-        if (ipActivity == null) {
-            // Means needed data is missing;  much clearer than NPE
-            String msg =
-                    "The following activity is not defined for owner '" + owner + "':  " + activity;
-            throw new RuntimeException(msg);
+        if (ipActivity != null) {
+            final IPermissionTargetProvider targetProvider =
+                    targetProviderRegistry.getTargetProvider(ipActivity.getTargetProviderKey());
+            final IPermissionTarget ipTarget = targetProvider.getTarget(target);
+            rslt =
+                    policy.doesPrincipalHavePermission(
+                            this, principal, ipOwner, ipActivity, ipTarget);
+        } else {
+            /*
+             * This circumstance means that a piece of the fundamental Permissions data expected by
+             * the code is missing in the database.  It normally happens when a newer version of the
+             * uPortal code is run against an existing database, and a required data update was
+             * overlooked.  This condition is not great, but probably not catastrophic;  it means
+             * that no one will (or can) have the new permission.  This method returns false.
+             *
+             * Administrators, however, have permission to do anything, including this unknown
+             * activity.  It's most common in uPortal for only Administrators to have access to
+             * exotic activities, so in most cases this omission is a wash.
+             *
+             * We need to log a WARNing, but this method is invoked a lot, and we don't want to do
+             * it incessantly.
+             */
+            final Long now = System.currentTimeMillis();
+            final String missingDataTrackerKey = owner + ":" + activity;
+            final Long lastLogMessageTime = missingDataLogTracker.get(missingDataTrackerKey);
+            if (lastLogMessageTime == null
+                    || lastLogMessageTime < now - MISSING_DATA_LOG_PERIOD_MILLIS) {
```

Comment:
Non-blocker to merging:
I still think that having a clearly defined data validation layer would be better for uPortal long term.

Without a validation layer, the business logic, in places like this, becomes more complex trying to handle the expected data flow, as well as a variety of acceptable unexpected flows.
And gets work-arounds/hacks like the `MISSING_DATA_LOG_PERIOD_MILLIS` conditional, which are easy to add, but will be hard to remove.